Common/GTesting/elxImpactMetricGTest.cxx

```diff
@@ -0,0 +1,190 @@
+  ExpectVectorEqual(GetVectorFromString<float>("", 42.0f), { 42.0f });
+  ExpectVectorEqual(GetVectorFromString<float>(4, "3.14", 0.0f), { 3.14f, 3.14f, 3.14f, 3.14f });
+  EXPECT_THROW(GetVectorFromString<float>("1e1000", 0.0f), itk::ExceptionObject);  // overflow
+  EXPECT_THROW(GetVectorFromString<float>("-1e1000", 0.0f), itk::ExceptionObject); // underflow
+  EXPECT_THROW(GetVectorFromString<float>("nan abc", 0.0f), itk::ExceptionObject); // parse fail
+  EXPECT_THROW(GetVectorFromString<float>("hello", 0.0f), itk::ExceptionObject);   // not numeric
+
+  // ---------- bool ----------
+  ExpectVectorEqual(GetVectorFromString<bool>("1 0", false), { true, false });
+  ExpectVectorEqual(GetBooleanVectorFromString("10", false), { true, false });
+  EXPECT_THROW(GetVectorFromString<bool>("maybe", false), itk::ExceptionObject); // parse fail
+  EXPECT_THROW(GetVectorFromString<bool>("10", false), itk::ExceptionObject);    // not space-separated
+}
+
+GTEST_TEST(groupStrByDimensions, Basic)
+{
+  std::vector<std::string> resultVec = groupStrByDimensions("5 5 5 8 6", { 3, 3 });
+
+  EXPECT_EQ(resultVec.size(), 2);
+  EXPECT_EQ(resultVec[0], "5 5 5");
+  EXPECT_EQ(resultVec[1], "8 6");
+
+  resultVec = groupStrByDimensions("5 5", { 3, 3 });
+
+  EXPECT_EQ(resultVec.size(), 2);
+  EXPECT_EQ(resultVec[0], "5 5");
+  EXPECT_EQ(resultVec[1], "");
+
+  resultVec = groupStrByDimensions("5 5 6 8 9 10 11 12", { 3, 3 });
+
+  EXPECT_EQ(resultVec.size(), 2);
+  EXPECT_EQ(resultVec[0], "5 5 6");
+  EXPECT_EQ(resultVec[1], "8 9 10");
+}
+
+GTEST_TEST(FormatParameterStringByDimensionAndLevel, Basic)
+{
+  // Create and initialize the elastix configuration
+  auto                                               config = elastix::Configuration::New();
+  elastix::Configuration::CommandLineArgumentMapType argMap;
+  itk::ParameterFileParser::ParameterMapType         parameterMap;
+
+  // --- First case: PatchSize as unsigned int ---
+
+  // Set PatchSize for 3 resolution levels
+  parameterMap["ImpactPatchSize"] = { "5", "5", "5", "11 11 11 29 29", "13", "13" };
+  config->Initialize(argMap, parameterMap);
+
+  // Level 0: read 3 individual values → expect "5 5 5"
+  std::string resultStr = formatParameterStringByDimensionAndLevel(config.GetPointer(), "Impact", "PatchSize", 0, 3);
+  EXPECT_EQ(resultStr, "5 5 5");
+
+  // Level 1: one line with 5 values → "11 11 11 29 29"
+  resultStr = formatParameterStringByDimensionAndLevel(config.GetPointer(), "Impact", "PatchSize", 1, 3);
+  EXPECT_EQ(resultStr, "11 11 11 29 29");
+
+  // Level 2: only "13" twice → expect fallback to fill 3 values
+  resultStr = formatParameterStringByDimensionAndLevel(config.GetPointer(), "Impact", "PatchSize", 2, 3);
+  EXPECT_EQ(resultStr, "13 13");
+
+  // --- Second case: PatchSize as float ---
+
+  parameterMap["ImpactPatchSize"] = { "6.0", "6.0", "3.0 3.0 1.5 1.5 1.5", "1.0", "1.0" };
+  config->Initialize(argMap, parameterMap);
```

Comment:
@vboussot Do I understand correctly that "PatchSize" only has whole numbers, by definition? (Looking at the METHOD section of your paper, "Each patch has a size _P_...") So in practice, it doesn't have floating point, right?

Reply:
Thanks @vboussot By the way, the METHOD section says:
![IMPACT-P](https://github.com/user-attachments/assets/83fbcede-f931-4068-a538-9cc5b033c81a)

which would suggest that for each _i_ in {1, ..., _D_} , _Pi_ is itself multi-dimensional (element of N^_D_). Is that OK? Maybe a typo 🤷